## __Recommender Systems in Python 101__
## __using Articles Sharing and Reading from CI&T Deskdrop Data__

In [2]:
import numpy as np
import scipy
import pandas as pd
import math
import random
import sklearn
from nltk.corpus import stopwords
from scipy.sparse import csr_matrix
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from scipy.sparse.linalg import svds
from sklearn.preprocessing import MinMaxScaler
import matplotlib.pyplot as plt

## __load data - CI&T Deskdrop__

In this section, we load the Deskdrop dataset, which contains a real sample of 12 months logs (Mar. 2016 - Feb. 2017) from CI&T's Internal Communication platform (DeskDrop). It contains about 73k logged users interactions on more than 3k public articles shared in the platform. It is composed of two CSV files:

shared_articles.csv
users_interactions.csv
Take a look in this kernels for a better picture of the dataset:

Deskdrop datasets EDA
DeskDrop Articles Topic Modeling

### __shared_articles.csv__

There are two possible event types at a given timestamp:

CONTENT SHARED: The article was shared in the platform and is available for users.
CONTENT REMOVED: The article was removed from the platform and not available for further recommendation.

In [3]:
# shared_articles.csv
articles_df = pd.read_csv('./input/shared_articles.csv')
articles_df = articles_df[articles_df['eventType'] == 'CONTENT SHARED']
articles_df.head(5)

timestamp       eventType            contentId       authorPersonId  \
1  1459193988  CONTENT SHARED -4110354420726924665  4340306774493623681   
2  1459194146  CONTENT SHARED -7292285110016212249  4340306774493623681   
3  1459194474  CONTENT SHARED -6151852268067518688  3891637997717104548   
4  1459194497  CONTENT SHARED  2448026894306402386  4340306774493623681   
5  1459194522  CONTENT SHARED -2826566343807132236  4340306774493623681   

       authorSessionId authorUserAgent authorRegion authorCountry contentType  \
1  8940341205206233829             NaN          NaN           NaN        HTML   
2  8940341205206233829             NaN          NaN           NaN        HTML   
3 -1457532940883382585             NaN          NaN           NaN        HTML   
4  8940341205206233829             NaN          NaN           NaN        HTML   
5  8940341205206233829             NaN          NaN           NaN        HTML   

                                                 url  \
1  http://www.nytimes.com/2016/03/28/business/dea...   
2  http://cointelegraph.com/news/bitcoin-future-w...   
3  https://cloudplatform.googleblog.com/2016/03/G...   
4  https://bitcoinmagazine.com/articles/ibm-wants...   
5  http://www.coindesk.com/ieee-blockchain-oxford...   

                                               title  \
1  Ethereum, a Virtual Currency, Enables Transact...   
2  Bitcoin Future: When GBPcoin of Branson Wins O...   
3                       Google Data Center 360° Tour   
4  IBM Wants to "Evolve the Internet" With Blockc...   
5  IEEE to Talk Blockchain at Cloud Computing Oxf...   

                                                text lang  
1  All of this work is still very early. The firs...   en  
2  The alarm clock wakes me at 8:00 with stream o...   en  
3  We're excited to share the Google Data Center ...   en  
4  The Aite Group projects the blockchain market ...   en  
5  One of the largest and oldest organizations fo...   en

In [4]:
articles_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3047 entries, 1 to 3121
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   timestamp        3047 non-null   int64 
 1   eventType        3047 non-null   object
 2   contentId        3047 non-null   int64 
 3   authorPersonId   3047 non-null   int64 
 4   authorSessionId  3047 non-null   int64 
 5   authorUserAgent  669 non-null    object
 6   authorRegion     669 non-null    object
 7   authorCountry    669 non-null    object
 8   contentType      3047 non-null   object
 9   url              3047 non-null   object
 10  title            3047 non-null   object
 11  text             3047 non-null   object
 12  lang             3047 non-null   object
dtypes: int64(4), object(9)
memory usage: 333.3+ KB


In [5]:
articles_df.columns

Index(['timestamp', 'eventType', 'contentId', 'authorPersonId',
       'authorSessionId', 'authorUserAgent', 'authorRegion', 'authorCountry',
       'contentType', 'url', 'title', 'text', 'lang'],
      dtype='object')

### __users_interactions.csv__

Contains logs of user interactions on shared articles. It can be joined to articles_shared.csv by contentId column.

The eventType values are:

VIEW: The user has opened the article.
LIKE: The user has liked the article.
COMMENT CREATED: The user created a comment in the article.
FOLLOW: The user chose to be notified on any new comment in the article.
BOOKMARK: The user has bookmarked the article for easy return in the future.

In [6]:
interactions_df = pd.read_csv('./input/users_interactions.csv')
interactions_df.head()

timestamp eventType            contentId             personId  \
0  1465413032      VIEW -3499919498720038879 -8845298781299428018   
1  1465412560      VIEW  8890720798209849691 -1032019229384696495   
2  1465416190      VIEW   310515487419366995 -1130272294246983140   
3  1465413895    FOLLOW   310515487419366995   344280948527967603   
4  1465412290      VIEW -7820640624231356730  -445337111692715325   

             sessionId                                          userAgent  \
0  1264196770339959068                                                NaN   
1  3621737643587579081  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2...   
2  2631864456530402479                                                NaN   
3 -3167637573980064150                                                NaN   
4  5611481178424124714                                                NaN   

  userRegion userCountry  
0        NaN         NaN  
1         NY          US  
2        NaN         NaN  
3        NaN         NaN  
4        NaN         NaN

In [7]:
interactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72312 entries, 0 to 72311
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   timestamp    72312 non-null  int64 
 1   eventType    72312 non-null  object
 2   contentId    72312 non-null  int64 
 3   personId     72312 non-null  int64 
 4   sessionId    72312 non-null  int64 
 5   userAgent    56918 non-null  object
 6   userRegion   56907 non-null  object
 7   userCountry  56918 non-null  object
dtypes: int64(4), object(4)
memory usage: 4.4+ MB


In [8]:
interactions_df.columns

Index(['timestamp', 'eventType', 'contentId', 'personId', 'sessionId',
       'userAgent', 'userRegion', 'userCountry'],
      dtype='object')

### __Data munging__

As there are different interactions types, we associate them with a weight or strength, assuming that, for example, a comment in an article indicates a higher interest of the user on the item than a like, or than a simple view.

In [9]:
event_type_strength = {
    'VIEW': 1.0,
    'LIKE': 2.0, 
    'BOOKMARK': 2.5, 
    'FOLLOW': 3.0,
    'COMMENT CREATED': 4.0,  
}

interactions_df['eventStrength'] = interactions_df['eventType'].apply\
(lambda x: event_type_strength[x])
interactions_df.head(5)

timestamp eventType            contentId             personId  \
0  1465413032      VIEW -3499919498720038879 -8845298781299428018   
1  1465412560      VIEW  8890720798209849691 -1032019229384696495   
2  1465416190      VIEW   310515487419366995 -1130272294246983140   
3  1465413895    FOLLOW   310515487419366995   344280948527967603   
4  1465412290      VIEW -7820640624231356730  -445337111692715325   

             sessionId                                          userAgent  \
0  1264196770339959068                                                NaN   
1  3621737643587579081  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2...   
2  2631864456530402479                                                NaN   
3 -3167637573980064150                                                NaN   
4  5611481178424124714                                                NaN   

  userRegion userCountry  eventStrength  
0        NaN         NaN            1.0  
1         NY          US            1.0  
2        NaN         NaN            1.0  
3        NaN         NaN            3.0  
4        NaN         NaN            1.0

In [10]:
interactions_df['eventStrength'].describe()

count    72312.000000
mean         1.236289
std          0.615299
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          4.000000
Name: eventStrength, dtype: float64

Recommender systems have a problem known as user cold-start, in which is hard do provide personalized recommendations for users with none or a very few number of consumed items, due to the lack of information to model their preferences.
For this reason, we are keeping in the dataset only users with at leas 5 interactions.

In [11]:
users_interactions_count_df = interactions_df.groupby(['personId', 'contentId']).size().groupby('personId').size()
print('# users: %d' % len(users_interactions_count_df))
users_interactions_count_df

# users: 1895


personId
-9223121837663643404    43
-9212075797126931087     5
-9207251133131336884     7
-9199575329909162940    11
-9196668942822132778     7
                        ..
 9165571805999894845     5
 9187866633451383747    17
 9191849144618614467    12
 9199170757466086545    14
 9210530975708218054    63
Length: 1895, dtype: int64

In [12]:
users_with_enough_interactions_df = users_interactions_count_df[users_interactions_count_df >= 5].reset_index()[['personId']]

In [13]:
print('# users with at least 5 interactions: %d' % len(users_with_enough_interactions_df))
users_with_enough_interactions_df

# users with at least 5 interactions: 1140


personId
0    -9223121837663643404
1    -9212075797126931087
2    -9207251133131336884
3    -9199575329909162940
4    -9196668942822132778
...                   ...
1135  9165571805999894845
1136  9187866633451383747
1137  9191849144618614467
1138  9199170757466086545
1139  9210530975708218054

[1140 rows x 1 columns]

In [14]:
print('# of interactions: %d' % len(interactions_df))

interactions_from_selected_users_df = interactions_df.merge(users_with_enough_interactions_df,
                                                           how = 'right',
                                                           left_on = 'personId',
                                                           right_on = 'personId')

print('# of interactions from users with at least 5 interactions: %d' % len(interactions_from_selected_users_df))
interactions_from_selected_users_df.head(5)

# of interactions: 72312
# of interactions from users with at least 5 interactions: 69868


timestamp eventType            contentId             personId  \
0  1463138398      VIEW  7516228655554309785 -9223121837663643404   
1  1463656314      VIEW  3041906492387035176 -9223121837663643404   
2  1464113091      VIEW -3750879736572068916 -9223121837663643404   
3  1462283851      VIEW  -730957269757756529 -9223121837663643404   
4  1462452127      VIEW -8949113594875411859 -9223121837663643404   

             sessionId                                          userAgent  \
0 -4482197405545551645  Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKi...   
1 -7824685088995468735  Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKi...   
2 -2774275024909061125  Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKi...   
3  2625340673871268120  Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKi...   
4 -3673331845456357462  Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKi...   

  userRegion userCountry  eventStrength  
0         SP          BR            1.0  
1         SP          BR            1.0  
2         SP          BR            1.0  
3         SP          BR            1.0  
4         SP          BR            1.0

In Deskdrop, users are allowed to view an article many times, and interact with them in different ways (eg. like or comment). Thus, to model the user interest on a given article, we aggregate all the interactions the user has performed in an item by a weighted sum of interaction type strength and apply a log transformation to smooth the distribution.

In [15]:
def smooth_user_preference(x):
    return math.log(1+x, 2)

interactions_full_df = interactions_from_selected_users_df\
    .groupby(['personId', 'contentId'])['eventStrength'].sum()\
    .apply(smooth_user_preference).reset_index()

print('# of unique user/item interactions: %d' % len(interactions_full_df))
interactions_full_df.head(10)

# of unique user/item interactions: 39106


personId            contentId  eventStrength
0 -9223121837663643404 -8949113594875411859       1.000000
1 -9223121837663643404 -8377626164558006982       1.000000
2 -9223121837663643404 -8208801367848627943       1.000000
3 -9223121837663643404 -8187220755213888616       1.000000
4 -9223121837663643404 -7423191370472335463       3.169925
5 -9223121837663643404 -7331393944609614247       1.000000
6 -9223121837663643404 -6872546942144599345       1.000000
7 -9223121837663643404 -6728844082024523434       1.000000
8 -9223121837663643404 -6590819806697898649       1.000000
9 -9223121837663643404 -6558712014192834002       1.584963

## __Evaluation__

Evaluation is important for machine learning projects, because it allows to compare objectivelly different algorithms and hyperparameter choices for models.
One key aspect of evaluation is to ensure that the trained model generalizes for data it was not trained on, using Cross-validation techniques. We are using here a simple cross-validation approach named holdout, in which a random data sample (20% in this case) are kept aside in the training process, and exclusively used for evaluation. All evaluation metrics reported here are computed using the test set.

Ps. A more robust evaluation approach could be to split train and test sets by a reference date, where the train set is composed by all interactions before that date, and the test set are interactions after that date. For the sake of simplicity, we chose the first random approach for this notebook, but you may want to try the second approach to better simulate how the recsys would perform in production predicting "future" users interactions.

In [16]:
interactions_train_df, interactions_test_df = train_test_split(interactions_full_df,
                                                             stratify=interactions_full_df['personId'],
                                                             test_size=0.20,
                                                             random_state=42)

print('# interactions on Train set: %d' % len(interactions_train_df))
print('# interactions on Test set: %d' % len(interactions_test_df))

# interactions on Train set: 31284
# interactions on Test set: 7822


In Recommender Systems, there are a set metrics commonly used for evaluation. We chose to work with Top-N accuracy metrics, which evaluates the accuracy of the top recommendations provided to a user, comparing to the items the user has actually interacted in test set.
This evaluation method works as follows:

- For each user
    - For each item the user has interacted in test set
        - Sample 100 other items the user has never interacted.
Ps. Here we naively assume those non interacted items are not relevant to the user, which might not be true, as the user may simply not be aware of those not interacted items. But let's keep this assumption.
        - Ask the recommender model to produce a ranked list of recommended items, from a set composed one interacted item and the 100 non-interacted ("non-relevant!) items
        - Compute the Top-N accuracy metrics for this user and interacted item from the recommendations ranked list
- Aggregate the global Top-N accuracy metrics


The Top-N accuracy metric choosen was Recall@N which evaluates whether the interacted item is among the top N items (hit) in the ranked list of 101 recommendations for a user.
Ps. Other popular ranking metrics are NDCG@N and MAP@N, whose score calculation takes into account the position of the relevant item in the ranked list (max. value if relevant item is in the first position). You can find a reference to implement this metrics in this post.

In [17]:
interactions_full_df

personId            contentId  eventStrength
0     -9223121837663643404 -8949113594875411859       1.000000
1     -9223121837663643404 -8377626164558006982       1.000000
2     -9223121837663643404 -8208801367848627943       1.000000
3     -9223121837663643404 -8187220755213888616       1.000000
4     -9223121837663643404 -7423191370472335463       3.169925
...                    ...                  ...            ...
39101  9210530975708218054  8477804012624580461       3.247928
39102  9210530975708218054  8526042588044002101       1.000000
39103  9210530975708218054  8856169137131817223       1.000000
39104  9210530975708218054  8869347744613364434       1.000000
39105  9210530975708218054  9209886322932807692       1.000000

[39106 rows x 3 columns]

In [18]:
#Indexing by personId to speed up the searches during evaluation
#무슨 의미인지 다시 파악
interactions_full_indexed_df = interactions_full_df.set_index('personId')
interactions_train_indexed_df = interactions_train_df.set_index('personId')
interactions_test_indexed_df = interactions_test_df.set_index('personId')

In [19]:
interactions_full_indexed_df

contentId  eventStrength
personId                                                
-9223121837663643404 -8949113594875411859       1.000000
-9223121837663643404 -8377626164558006982       1.000000
-9223121837663643404 -8208801367848627943       1.000000
-9223121837663643404 -8187220755213888616       1.000000
-9223121837663643404 -7423191370472335463       3.169925
...                                   ...            ...
 9210530975708218054  8477804012624580461       3.247928
 9210530975708218054  8526042588044002101       1.000000
 9210530975708218054  8856169137131817223       1.000000
 9210530975708218054  8869347744613364434       1.000000
 9210530975708218054  9209886322932807692       1.000000

[39106 rows x 2 columns]

In [20]:
def get_items_interacted(person_id, interactions_df):
    # Get the user's data and merge in the movie information.
    interacted_items = interactions_df.loc[person_id]['contentId']
    
    return set(interacted_items if type(interacted_items) == pd.Series else [interacted_items])

In [21]:
#Top-N accuracy metrics consts
EVAL_RANDOM_SAMPLE_NON_INTERACTED_ITEMS = 100

class ModelEvaluator:


    def get_not_interacted_items_sample(self, person_id, sample_size, seed=42):
        interacted_items = get_items_interacted(person_id, interactions_full_indexed_df)
        all_items = set(articles_df['contentId'])
        non_interacted_items = all_items - interacted_items

        random.seed(seed)
        non_interacted_items_sample = random.sample(non_interacted_items, sample_size)
        return set(non_interacted_items_sample)

    def _verify_hit_top_n(self, item_id, recommended_items, topn):        
            try:
                index = next(i for i, c in enumerate(recommended_items) if c == item_id)
            except:
                index = -1
            hit = int(index in range(0, topn))
            return hit, index

    def evaluate_model_for_user(self, model, person_id):
        #Getting the items in test set
        interacted_values_testset = interactions_test_indexed_df.loc[person_id]
        if type(interacted_values_testset['contentId']) == pd.Series:
            person_interacted_items_testset = set(interacted_values_testset['contentId'])
        else:
            person_interacted_items_testset = set([int(interacted_values_testset['contentId'])])  
        interacted_items_count_testset = len(person_interacted_items_testset) 

        #Getting a ranked recommendation list from a model for a given user
        person_recs_df = model.recommend_items(person_id, 
                                               items_to_ignore=get_items_interacted(person_id, 
                                                                                    interactions_train_indexed_df), 
                                               topn=10000000000)

        hits_at_5_count = 0
        hits_at_10_count = 0
        #For each item the user has interacted in test set
        for item_id in person_interacted_items_testset:
            #Getting a random sample (100) items the user has not interacted 
            #(to represent items that are assumed to be no relevant to the user)
            non_interacted_items_sample = self.get_not_interacted_items_sample(person_id, 
                                                                          sample_size=EVAL_RANDOM_SAMPLE_NON_INTERACTED_ITEMS, 
                                                                          seed=item_id%(2**32))

            #Combining the current interacted item with the 100 random items
            items_to_filter_recs = non_interacted_items_sample.union(set([item_id]))

            #Filtering only recommendations that are either the interacted item or from a random sample of 100 non-interacted items
            valid_recs_df = person_recs_df[person_recs_df['contentId'].isin(items_to_filter_recs)]                    
            valid_recs = valid_recs_df['contentId'].values
            #Verifying if the current interacted item is among the Top-N recommended items
            hit_at_5, index_at_5 = self._verify_hit_top_n(item_id, valid_recs, 5)
            hits_at_5_count += hit_at_5
            hit_at_10, index_at_10 = self._verify_hit_top_n(item_id, valid_recs, 10)
            hits_at_10_count += hit_at_10

        #Recall is the rate of the interacted items that are ranked among the Top-N recommended items, 
        #when mixed with a set of non-relevant items
        recall_at_5 = hits_at_5_count / float(interacted_items_count_testset)
        recall_at_10 = hits_at_10_count / float(interacted_items_count_testset)

        person_metrics = {'hits@5_count':hits_at_5_count, 
                          'hits@10_count':hits_at_10_count, 
                          'interacted_count': interacted_items_count_testset,
                          'recall@5': recall_at_5,
                          'recall@10': recall_at_10}
        return person_metrics

    def evaluate_model(self, model):
        #print('Running evaluation for users')
        people_metrics = []
        for idx, person_id in enumerate(list(interactions_test_indexed_df.index.unique().values)):
            #if idx % 100 == 0 and idx > 0:
            #    print('%d users processed' % idx)
            person_metrics = self.evaluate_model_for_user(model, person_id)  
            person_metrics['_person_id'] = person_id
            people_metrics.append(person_metrics)
        print('%d users processed' % idx)

        detailed_results_df = pd.DataFrame(people_metrics) \
                            .sort_values('interacted_count', ascending=False)
        
        global_recall_at_5 = detailed_results_df['hits@5_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        global_recall_at_10 = detailed_results_df['hits@10_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        
        global_metrics = {'modelName': model.get_model_name(),
                          'recall@5': global_recall_at_5,
                          'recall@10': global_recall_at_10}    
        return global_metrics, detailed_results_df
    
model_evaluator = ModelEvaluator()

## __Popularity model__

A common (and usually hard-to-beat) baseline approach is the Popularity model. This model is not actually personalized - it simply recommends to a user the most popular items that the user has not previously consumed. As the popularity accounts for the "wisdom of the crowds", it usually provides good recommendations, generally interesting for most people.
Ps. The main objective of a recommender system is to leverage the long-tail items to the users with very specific interests, which goes far beyond this simple technique.

In [22]:
# Compute the most popular items
item_popularity_df = interactions_full_df.groupby('contentId')['eventStrength'].sum().sort_values(ascending=False).reset_index()
item_popularity_df.head()

contentId  eventStrength
0 -4029704725707465084     307.733799
1 -6783772548752091658     233.762157
2  -133139342397538859     228.024567
3 -8208801367848627943     197.107608
4 -6843047699859121724     193.825208

In [23]:
class PopularityRecommender:
    
    MODEL_NAME = 'Popularity'
    
    def __init__(self, popularity_df, items_df=None):
        self.popularity_df = popularity_df
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Recommend the more popular items that the user hasn't seen yet.
        recommendations_df = self.popularity_df[~self.popularity_df['contentId'].isin(items_to_ignore)] \
                               .sort_values('eventStrength', ascending = False) \
                               .head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['eventStrength', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df
    
popularity_model = PopularityRecommender(item_popularity_df, articles_df)

Here we perform the evaluation of the Popularity model, according to the method described above.

It achieved the Recall@5 of 0.2417, which means that about 24% of interacted items in test set were ranked by Popularity model among the top-5 items (from lists with 100 random items). And Recall@10 was even higher (37%), as expected.

It might be surprising to you that usually Popularity models could perform so well!

In [24]:
print('Evaluating Popularity recommendation model...')
pop_global_metrics, pop_detailed_results_df = model_evaluator.evaluate_model(popularity_model)
print('\nGlobal metrics:\n%s' % pop_global_metrics)
pop_detailed_results_df.head(10)

Evaluating Popularity recommendation model...
1139 users processed

Global metrics:
{'modelName': 'Popularity', 'recall@5': 0.2418818716440808, 'recall@10': 0.3725389925850166}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             28             50               192  0.145833   0.260417   
17             12             25               134  0.089552   0.186567   
16             13             23               130  0.100000   0.176923   
10              5              9               117  0.042735   0.076923   
82             26             40                88  0.295455   0.454545   
161            12             18                80  0.150000   0.225000   
65             20             34                73  0.273973   0.465753   
81             17             23                69  0.246377   0.333333   
106            14             18                69  0.202899   0.260870   
52             21             28                68  0.308824   0.411765   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

## __Content-based Filtering model__

Content-based filtering approaches leverage description or attributes from items the user has interacted to recommend similar items. It depends only on the user previous choices, making this method robust to avoid the cold-start problem. For textual items, like articles, news and books, it is simple to use the raw text to build item profiles and user profiles.

Here we are using a very popular technique in information retrieval (search engines) named TF-IDF. This technique converts unstructured text into a vector structure, where each word is represented by a position in the vector, and the value measures how relevant a given word is for an article. As all items will be represented in the same Vector Space Model, it is to compute similarity between articles.

See this presentation (from slide 30) for more information on TF-IDF and Cosine similarity.

In [25]:
#Ignoring stopword (words with no semantics) from English and Protuguese(as we have a corpus with mixed languages)
stopwords_list = stopwords.words('english') + stopwords.words('portuguese')

#Trains a model whose vercotrs size is 5000, composed by the main unigrams and bigrams found in the corpus, ignoring stopwords
vectorizer = TfidfVectorizer(analyzer='word',
                            ngram_range=(1, 2),
                            min_df=0.003,
                            max_df=0.5,
                            max_features=5000,
                            stop_words=stopwords_list)

item_ids = articles_df['contentId'].tolist()
tfidf_matrix = vectorizer.fit_transform(articles_df['title'] + "" + articles_df['text'])
tfidf_feature_names = vectorizer.get_feature_names()
tfidf_matrix

<3047x5000 sparse matrix of type '<class 'numpy.float64'>'
	with 638928 stored elements in Compressed Sparse Row format>

To model the user profile, we take all the item profiles the user has interacted and average them. The average is weighted by the interaction strength, in other words, the articles the user has interacted the most (eg. liked or commented) will have a higher strength in the final user profile.

In [26]:
def get_item_profile(item_id):
    idx = item_ids.index(item_id)
    item_profile = tfidf_matrix[idx:idx+1]
    return item_profile

def get_item_profiles(ids):
    item_profiles_list = [get_item_profile(x) for x in ids]
    item_profiles = scipy.sparse.vstack(item_profiles_list)
    return item_profiles

def build_users_profile(person_id, interactions_indexed_df):
    interactions_person_df = interactions_indexed_df.loc[person_id]
    user_item_profiles = get_item_profiles(interactions_person_df['contentId'])
    
    user_item_strengths = np.array(interactions_person_df['eventStrength']).reshape(-1,1)
    #Weighted average of item profiles by the interactions strength
    user_item_strengths_weighted_avg = np.sum(user_item_profiles.multiply(user_item_strengths), axis=0) / np.sum(user_item_strengths)
    user_profile_norm = sklearn.preprocessing.normalize(user_item_strengths_weighted_avg)
    return user_profile_norm

def build_users_profiles(): 
    interactions_indexed_df = interactions_train_df[interactions_train_df['contentId'] \
                                                   .isin(articles_df['contentId'])].set_index('personId')
    user_profiles = {}
    for person_id in interactions_indexed_df.index.unique():
        user_profiles[person_id] = build_users_profile(person_id, interactions_indexed_df)
    return user_profiles

In [27]:
user_profiles = build_users_profiles()
len(user_profiles)

1140

Let's take a look in the profile. It is a unit vector of 5000 length. The value in each position represents how relevant is a token (unigram or bigram) for me.

Looking my profile, it appears that the top relevant tokens really represent my professional interests in machine learning, deep learning, artificial intelligence and google cloud platform! So we might expect good recommendations here!

In [28]:
myprofile = user_profiles[-1479311724257856983]
print(myprofile.shape)
pd.DataFrame(sorted(zip(tfidf_feature_names, 
                        user_profiles[-1479311724257856983].flatten().tolist()), key=lambda x: -x[1])[:15],
             columns=['token', 'relevance'])

(1, 5000)


token  relevance
0           learning   0.298732
1   machine learning   0.245992
2            machine   0.237843
3             google   0.202839
4               data   0.169776
5                 ai   0.156203
6         algorithms   0.115666
7               like   0.097744
8           language   0.087609
9             people   0.082024
10              deep   0.081542
11     deep learning   0.080979
12          research   0.076020
13         algorithm   0.074905
14             apple   0.074050

In [29]:
class ContentBasedRecommender:
    
    MODEL_NAME = 'Content-Based'
    
    def __init__(self, items_df=None):
        self.item_ids = item_ids
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def _get_similar_items_to_user_profile(self, person_id, topn=1000):
        #Computes the cosine similarity between the user profile and all item profiles
        cosine_similarities = cosine_similarity(user_profiles[person_id], tfidf_matrix)
        #Gets the top similar items
        similar_indices = cosine_similarities.argsort().flatten()[-topn:]
        #Sort the similar items by similarity
        similar_items = sorted([(item_ids[i], cosine_similarities[0,i]) for i in similar_indices], key=lambda x: -x[1])
        return similar_items
    
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        similar_items = self._get_similar_items_to_user_profile(user_id)
        #Ignore items the user has already interacted
        similar_items_filtered = list(filter(lambda x: x[0] not in items_to_ignore, similar_items))
        
        recommendations_df = pd.DataFrame(similar_items_filtered, columns=['contentId', 'recStrength']).head(topn)
        
        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')
                
            recommendations_df = recommendations_df.merge(self.items_df, how='left', left_on='contentId',
                                                         right_on='contentId')[['recStrength', 'contentId', 'title', 'url', 'lang']]
        return recommendations_df
    
content_based_recommender_model = ContentBasedRecommender(articles_df)

With personalized recommendations of content-based filtering model, we have a Recall@5 to about 0.162, which means that about 16% of interacted items in test set were ranked by this model among the top-5 items (from lists with 100 random items). And Recall@10 was 0.261 (52%). The lower performance of the Content-Based model compared to the Popularity model may indicate that users are not that fixed in content very similar to their previous reads.

In [30]:
print('Evaluating Content-Based Filtering model...')
cb_global_metrics, cb_detailed_results_df = model_evaluator.evaluate_model(content_based_recommender_model)
print("\nGlobal metrics: \n%s" % cb_global_metrics)
cb_detailed_results_df.head(10)

Evaluating Content-Based Filtering model...
1139 users processed

Global metrics: 
{'modelName': 'Content-Based', 'recall@5': 0.16287394528253643, 'recall@10': 0.2614420864229097}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             15             24               192  0.078125   0.125000   
17             18             29               134  0.134328   0.216418   
16             20             33               130  0.153846   0.253846   
10             32             47               117  0.273504   0.401709   
82              6             15                88  0.068182   0.170455   
161            11             23                80  0.137500   0.287500   
65              8             13                73  0.109589   0.178082   
81              8             19                69  0.115942   0.275362   
106             3              9                69  0.043478   0.130435   
52              3              8                68  0.044118   0.117647   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

## __Collaborative Filtering Model__

Collaborative Filtering(CF) has two main implementation stratgies:

- __Memory-based__: This approach uses the memory of previous users interactions to compute users similarities based on items they've interacted (user-based approach) or compute items similarities based on the users that have interacted with them (item-based approach).
    
    A typical example of this approach is User Neighbourhood-based CF, in which the top-N similar users(usually computed using Pearson correlation) for a user are seleceted and used to recommend items those similar users liked, but the current user have not interacted yet. This approach is very simple to implement, but usually do not scale well for many users. A nice Python implementation the approach in availablein Crab.

- __Model-based__: this approach, models are developed using different machine learning algorithms to recommend items to users. There are many model-based CF algorithms, like neural networks, bayesian networks, clustering model, and latent factor models such as Singular Value Decomposition(SVD) and, probabilistic latent semantic analysis.

### __Matrix Factorization(MF)__

Latent factor models compress user-item matrix into a low-dimensional representation in terms of latent factors. One advantage of using this approach is that instead of having a high dimensional matrix containing abundant number of missing values we will be dealing with a much smaller matrix in lower-dimensional space. A reduced presentation could be utillzed for either user-based or item-based neighborhood algorithms that are presented in the previous section. There are several avantages with this paradigm. It handles the sparsity of the original matrix better than memory based ones. Also comparing similarity on the resulting matrix is much more scalable especially in dealing with large sparse datasets.


Here we a user popular latent factor model named Singulaar Value Decomposition(SVD). There are other matrix factorization frameworks more specific to CF you might try, like surprise, mrec or python-recsys. We chose a SciPy implemenation of SVD because it is available on Kaggle kernels. P.s, See an example of SVD on a movies dataset in the blog post.


An important decision is the number of factors to factor the user-item matrix. The higher the number of factors, the more precise is t he factorization in the original matrix reconstructions. Therefore, if the model is allowed to memorize too much details of the original matrix, it may not generalize well for data if was not trained on, Reducing the number of factors increases the model generalization.




In [31]:
#Creating a sparse pivot table with users in rows and items in columns
users_items_pivot_matrix_df = interactions_train_df.pivot(index='personId',
                                                         columns='contentId',
                                                         values='eventStrength').fillna(0)

users_items_pivot_matrix_df.head(10)

contentId             -9222795471790223670  -9216926795620865886  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   2.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9194572880052200111  -9192549002213406534  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9190737901804729417  -9189659052158407108  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9176143510534135851  -9172673334835262304  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9171475473795142532  -9166778629773133902  ...  \
personId                                                          ...   
-9223121837663643404                   0.0                   0.0  ...   
-9212075797126931087                   0.0                   0.0  ...   
-9207251133131336884                   0.0                   0.0  ...   
-9199575329909162940                   0.0                   0.0  ...   
-9196668942822132778                   0.0                   0.0  ...   
-9188188261933657343                   0.0                   0.0  ...   
-9172914609055320039                   0.0                   0.0  ...   
-9156344805277471150                   0.0                   0.0  ...   
-9

In [32]:
users_items_pivot_matrix = users_items_pivot_matrix_df.values
users_items_pivot_matrix[:10]

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 2., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [33]:
users_ids = list(users_items_pivot_matrix_df.index)
users_ids[:10]

[-9223121837663643404,
 -9212075797126931087,
 -9207251133131336884,
 -9199575329909162940,
 -9196668942822132778,
 -9188188261933657343,
 -9172914609055320039,
 -9156344805277471150,
 -9120685872592674274,
 -9109785559521267180]

In [34]:
users_items_pivot_sparse_matrix = csr_matrix(users_items_pivot_matrix)
users_items_pivot_sparse_matrix

<1140x2926 sparse matrix of type '<class 'numpy.float64'>'
	with 31284 stored elements in Compressed Sparse Row format>

In [35]:
#The number of factors to factor the user-item matrix
NUMBER_OF_FACTORS_MF = 15
#Performs matrix factorization of the original user item matrix
#U, sigman, Vt = svds(users_items_pivot_matrix, k=NUMBER_OF_FACTORS_MF)
U, sigma, Vt = svds(users_items_pivot_sparse_matrix, k = NUMBER_OF_FACTORS_MF)

In [36]:
U.shape

(1140, 15)

In [37]:
Vt.shape

(15, 2926)

In [38]:
sigma = np.diag(sigma)
sigma.shape

(15, 15)

After the factorization, we try to reconstruct the the original matrix by multiplying its factors. The resulting matrix is not sparse any more. It was generated predictions for items the user have not yet interaction, which we will exploit for recommenddations.

In [39]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt)
all_user_predicted_ratings

array([[ 0.01039915,  0.00081872, -0.01725263, ...,  0.00140708,
         0.0110647 ,  0.00226063],
       [-0.00019285, -0.00031318, -0.00264624, ...,  0.00251658,
         0.00017609, -0.00189488],
       [-0.01254721,  0.0065947 , -0.00590676, ...,  0.00698975,
        -0.01015696,  0.01154572],
       ...,
       [-0.02995379,  0.00805715, -0.01846307, ..., -0.01083078,
        -0.00118591,  0.0096798 ],
       [-0.01845505,  0.00467019,  0.01219602, ...,  0.00409507,
         0.00019482, -0.00752562],
       [-0.01506374,  0.00327732,  0.13391269, ..., -0.01191815,
         0.06422074,  0.01303244]])

In [40]:
all_user_predicted_ratings_norm = (all_user_predicted_ratings - all_user_predicted_ratings.min() /
                                 (all_user_predicted_ratings.max() - all_user_predicted_ratings.min()))

In [41]:
#Converting the reconstructed matrix back to a Pandas dataframe
cf_preds_df = pd.DataFrame(all_user_predicted_ratings_norm, columns=users_items_pivot_matrix_df.columns,
                          index=users_ids).transpose()
cf_preds_df.head(10)

-9223121837663643404  -9212075797126931087  \
contentId                                                          
-9222795471790223670              0.148351              0.137759   
-9216926795620865886              0.138770              0.137638   
-9194572880052200111              0.120699              0.135305   
-9192549002213406534              0.173032              0.138343   
-9190737901804729417              0.157890              0.133152   
-9189659052158407108              0.146612              0.144556   
-9176143510534135851              0.184378              0.144326   
-9172673334835262304              0.143034              0.138569   
-9171475473795142532              0.162404              0.137191   
-9166778629773133902              0.147111              0.135953   

                      -9207251133131336884  -9199575329909162940  \
contentId                                                          
-9222795471790223670              0.125404              0.190910   
-9216926795620865886              0.144546              0.137302   
-9194572880052200111              0.132045              0.134285   
-9192549002213406534              0.108906              0.130250   
-9190737901804729417              0.144636              0.144314   
-9189659052158407108              0.139414              0.135034   
-9176143510534135851              0.151754              0.148216   
-9172673334835262304              0.140797              0.136848   
-9171475473795142532              0.138915              0.135134   
-9166778629773133902              0.125307              0.135931   

                      -9196668942822132778  -9188188261933657343  \
contentId                                                          
-9222795471790223670              0.127916              0.132539   
-9216926795620865886              0.138102              0.138290   
-9194572880052200111              0.159251              0.136673   
-9192549002213406534              0.155566              0.143448   
-9190737901804729417              0.135954              0.140308   
-9189659052158407108              0.146502              0.136315   
-9176143510534135851              0.135413              0.133499   
-9172673334835262304              0.138350              0.138706   
-9171475473795142532              0.140417              0.139166   
-9166778629773133902              0.143319              0.137947   

                      -9172914609055320039  -9156344805277471150  \
contentId                                                          
-9222795471790223670              0.134034              0.187253   
-9216926795620865886              0.138149              0.138812   
-9194572880052200111              0.167052              0.123221   
-9192549002213406534              0.151387              0.195514   
-9190737901804729417              0.141828              0.142069   
-9189659052158407108              0.143395              0.140144   
-9176143510534135851              0.154566              0.161021   
-9172673334835262304              0.139297              0.140665   
-9171475473795142532              0.147291              0.134422   
-9166778629773133902              0.140050              0.142757   

                      -9120685872592674274  -9109785559521267180  ...  \
contentId                                                         ...   
-9222795471790223670              0.124493              0.117900  ...   
-9216926795620865886              0.140293              0.139717  ...   
-9194572880052200111              0.114542              0.142901  ...   
-9192549002213406534              0.185483              0.142160  ...   
-9190737901804729417              0.110854              0.157745  ...   
-9189659052158407108              0.161099              0.159262  ...   
-9176143510534135851              0.197533              0.195258  ...   
-9172673334835262304              0.135140              0.141882  ...   
-9171475473795142532    

In [42]:
len(cf_preds_df.columns)

1140

In [43]:
class CFRecommender:
    
    MODEL_NAME = 'Collaborative Filtering'
    
    def __init__(self, cf_predictions_df, items_df=None):
        self.cf_predictions_df = cf_predictions_df
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Get and sort the user's predictions
        sorted_user_predictions = self.cf_predictions_df[user_id].sort_values(ascending=False) \
                                    .reset_index().rename(columns={user_id: 'recStrength'})

        # Recommend the highest predicted rating movies that the user hasn't seen yet.
        recommendations_df = sorted_user_predictions[~sorted_user_predictions['contentId'].isin(items_to_ignore)] \
                               .sort_values('recStrength', ascending = False) \
                               .head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['recStrength', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df
    
cf_recommender_model = CFRecommender(cf_preds_df, articles_df)

Evaluating the Collaborative Filtering model (SVD matrix factorization), we observe that we got Recall@5 (33%) and Recall@10 (46%) values, much higher than Popularity model and Content-Based model.

In [44]:
print('Evaluating Collaborative Filtering (SVD Matrix Factorization) model...')
cf_global_metrics, cf_detailed_results_df = model_evaluator.evaluate_model(cf_recommender_model)
print('\nGlobal metrics:\n%s' % cf_global_metrics)
cf_detailed_results_df.head(10)

Evaluating Collaborative Filtering (SVD Matrix Factorization) model...
1139 users processed

Global metrics:
{'modelName': 'Collaborative Filtering', 'recall@5': 0.33392994119151115, 'recall@10': 0.46803886474047557}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             21             46               192  0.109375   0.239583   
17             30             56               134  0.223881   0.417910   
16             16             34               130  0.123077   0.261538   
10             38             51               117  0.324786   0.435897   
82             39             48                88  0.443182   0.545455   
161            22             34                80  0.275000   0.425000   
65             24             32                73  0.328767   0.438356   
81             16             21                69  0.231884   0.304348   
106            20             28                69  0.289855   0.405797   
52             23             30                68  0.338235   0.441176   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

## __Hybrid Recommender__

What if we combine Collaborative Filtering and Content-Based Filtering approaches?<br>
Would that provideaaa us with more accurate recommendations?<br>
In fact, hybrid methods have performed better than individual approaches in many studies and have being extensively used by researchers and practioners.<br>
Let's build a simple hybridization method, as an ensemble that takes the weighted average of the normalized CF scores with the Content-Based scores, and ranking by resulting score. In this case, as the CF model is much more accurate than the CB model, the weights for the CF and CB models are 100.0 and 1.0, respectivelly.

In [50]:
class HybridRecommender:
    
    MODEL_NAME = 'Hybrid'
    
    def __init__(self, cb_rec_model, cf_rec_model, items_df, cb_ensemble_weight=1.0, cf_ensemble_weight=1.0):
        self.cb_rec_model = cb_rec_model
        self.cf_rec_model = cf_rec_model
        self.cb_ensemble_weight = cb_ensemble_weight
        self.cf_ensemble_weight = cf_ensemble_weight
        self.items_df = items_df
    
    
    def get_model_name(self):
        return self.MODEL_NAME
    
    
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        #Getting the top-1000 Content-based filtering recommendations
        cb_recs_df = self.cb_rec_model.recommend_items(user_id, items_to_ignore=items_to_ignore,
                                                       verbose=verbose, topn=1000).rename(columns={'recStrength': 'recStrengthCB'})
        
        #Getting the top-1000 Collaborative filtering recommendations
        cf_recs_df = self.cf_rec_model.recommend_items(user_id, items_to_ignore=items_to_ignore,
                                                       verbose=verbose, topn=1000).rename(columns={'recStrength': 'recStrengthCF'})
        
        #Combining the results by contentId
        recs_df = cb_recs_df.merge(cf_recs_df,
                                  how='outer',
                                  left_on='contentId',
                                  right_on='contentId').fillna(0.0)
        
        #Computing a hybrid recommendation score based on CF and CB scores
        #recs_df['recStrengthHybrid'] = recs_df['recStrengthCB'] * recs_df['recStrengthCF'] 
        recs_df['recStrengthHybrid'] = (recs_df['recStrengthCB'] * self.cb_ensemble_weight) 
        +(recs_df['recStrengthCF'] * self.cf_ensemble_weight)
        
        #Sorting recommendations by hybrid score
        recommendations_df = recs_df.sort_values('recStrengthHybrid', ascending=False).head(topn)
        
        if verbose:
            if self.items_df is None:
                raise Excecption('"items_df" is required in verbose mode')
                
            recommendations_df = recommendations_df.merge(self.items_df,
                                                         how='left',
                                                         left_on='contentId',
                                                         right_on='contentId')[['recStrengthHybrid', 'contentId', 'title', 'url', 'lang']]
            
        return recommendations_df

hybrid_recommender_model = HybridRecommender(content_based_recommender_model, cf_recommender_model, articles_df,
                                            cb_ensemble_weight=1.0, cf_ensemble_weight=100.0)

We have a new champion!<br>
Our simple hybrid approach surpasses Content-Based filtering with its combination with Collaborative Filtering. Now we have a Recall@5 of 34.2% and Recall@10 of 47.9%

In [51]:
print('Evaluating Hybrid model...')
hybrid_global_metrics, hybrid_detailed_results_df = model_evaluator.evaluate_model(hybrid_recommender_model)
print('\nGlobal metrics:\n%s' % hybrid_global_metrics)
hybrid_detailed_results_df.head(10)

Evaluating Hybrid model...
1139 users processed

Global metrics:
{'modelName': 'Hybrid', 'recall@5': 0.16287394528253643, 'recall@10': 0.2614420864229097}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             15             24               192  0.078125   0.125000   
17             18             29               134  0.134328   0.216418   
16             20             33               130  0.153846   0.253846   
10             32             47               117  0.273504   0.401709   
82              6             15                88  0.068182   0.170455   
161            11             23                80  0.137500   0.287500   
65              8             13                73  0.109589   0.178082   
81              8             19                69  0.115942   0.275362   
106             3              9                69  0.043478   0.130435   
52              3              8                68  0.044118   0.117647   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

### __Comparing the method__